test_pv_pool_with_nfs is frequently ( not all the time) failing on 4.21 and 4.20 - Investigate and fix if needed

test_pv_pool_with_nfs is frequently ( not all the time) failing on 4.21 and 4.20 . 

Relevant 4.20 runs: 

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/954/44298/2021718/2021905/log
https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/954/45046/2041954/2042082/log

Relevant 4.21 runs: 

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/43424/1995706/1995894/log
https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/43445/1996436/1996626/log 

etc

Error: 

E           AssertionError: NFS provisioner pods didn't start up successfully
E           assert False
E            +  where False = wait_for_pods_to_be_in_statuses(expected_statuses='Running', namespace='nfs-storage')
E            +    where 'Running' = constants.STATUS_RUNNING
E            +    and   'nfs-storage' = constants.NFS_NAMESPACE_NAME



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/conftest.py:10620: AssertionError

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -5,7 +5,6 @@ import os
 import pandas as pd
 import random
 import secrets
-import socket
 import string
 import time
 import tempfile
@@ -10771,12 +10770,28 @@ def setup_nfs(request, setup_rbac):
                 f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
             )
 
+        log.info(f"Checking DNS resolution of NFS server '{nfs_server}' from cluster")
         try:
-            socket.getaddrinfo(nfs_server, None, socket.AF_UNSPEC, socket.SOCK_STREAM)
-        except socket.gaierror:
-            pytest.skip(
-                f"NFS server '{nfs_server}' is not DNS-resolvable from this cluster; "
-                f"skipping NFS test"
+            node_name = (
+                exec_cmd("oc get nodes -o jsonpath='{.items[0].metadata.name}'")
+                .stdout.decode()
+                .strip()
+                .strip("'")
+            )
+            result = exec_cmd(
+                f"oc debug node/{node_name} --to-namespace=default "
+                f"-- chroot /host getent hosts {nfs_server}",
+                ignore_error=True,
+            )
+            if not result.stdout or not result.stdout.decode().strip():
+                pytest.skip(
+                    f"NFS server '{nfs_server}' is not DNS-resolvable from "
+                    f"cluster node {node_name}; skipping NFS test"
+                )
+        except Exception:
+            log.warning(
+                f"Could not verify DNS resolution of '{nfs_server}' from cluster, "
+                f"proceeding with test"
             )
 
         log.info(

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -10763,6 +10763,13 @@ def setup_nfs(request, setup_rbac):
 
         # Update the deployment with ocs-ci noobaa nfs server
         # noobaa nfs mount path and create the deployment
+        nfs_server = config.ENV_DATA.get("nb_nfs_server")
+        nfs_mount = config.ENV_DATA.get("nb_nfs_mount")
+        if not nfs_server or not nfs_mount:
+            raise ValueError(
+                f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
+            )
+
         log.info(
             "Updating NFS client provisioner data with Noobaa NFS server and NFS mount path info"
         )
@@ -10771,25 +10778,49 @@ def setup_nfs(request, setup_rbac):
         )
         nfs_client_prov_dep_data["spec"]["template"]["spec"]["containers"][0]["env"][1][
             "value"
-        ] = config.ENV_DATA.get("nb_nfs_server")
+        ] = nfs_server
         nfs_client_prov_dep_data["spec"]["template"]["spec"]["containers"][0]["env"][2][
             "value"
-        ] = config.ENV_DATA.get("nb_nfs_mount")
+        ] = nfs_mount
         nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"][0]["nfs"][
             "server"
-        ] = config.ENV_DATA.get("nb_nfs_server")
+        ] = nfs_server
         nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"][0]["nfs"][
             "path"
-        ] = config.ENV_DATA.get("nb_nfs_mount")
+        ] = nfs_mount
         log.info("Creating NFS client provisioner deployment")
         create_resource(**nfs_client_prov_dep_data)
         nfs_client_dep_obj = OCS(**nfs_client_prov_dep_data)
 
         log.info("Waiting for NFS client provisioner pods to be running")
-        assert wait_for_pods_to_be_in_statuses(
+        pods_running = wait_for_pods_to_be_in_statuses(
             expected_statuses=constants.STATUS_RUNNING,
             namespace=constants.NFS_NAMESPACE_NAME,
-        ), "NFS provisioner pods didn't start up successfully"
+            timeout=600,
+        )
+        if not pods_running:
+            ocp_pod = OCP(kind=constants.POD, namespace=constants.NFS_NAMESPACE_NAME)
+            try:
+                pod_data = ocp_pod.get()
+                for pod in pod_data.get("items", []):
+                    pod_name = pod["metadata"]["name"]
+                    phase = pod.get("status", {}).get("phase")
+                    log.error(f"NFS provisioner pod {pod_name} phase: {phase}")
+                    for cond in pod.get("status", {}).get("conditions", []):
+                        log.error(
+                            f"  Condition {cond.get('type')}: "
+                            f"{cond.get('status')} - {cond.get('message', '')}"
+                        )
+            except Exception:
+                log.exception("Failed to get NFS pod details for diagnostics")
+            try:
+                ocp_event = OCP(kind="Event", namespace=constants.NFS_NAMESPACE_NAME)
+                event_data = ocp_event.get()
+                for event in event_data.get("items", []):
+                    log.error(f"  Event: {event.get('reason')}: {event.get('message')}")
+            except Exception:
+                log.exception("Failed to get NFS namespace events for diagnostics")
+        assert pods_running, "NFS provisioner pods didn't start up successfully"
 
     finally:
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -10780,14 +10780,28 @@ def setup_nfs(request, setup_rbac):
             0
         ]
         env_by_name = {e.get("name"): e for e in container.get("env", [])}
+        missing_env = [k for k in ("NFS_SERVER", "NFS_PATH") if k not in env_by_name]
+        if missing_env:
+            raise ValueError(
+                f"NFS deployment template missing env vars: {', '.join(missing_env)}"
+            )
         env_by_name["NFS_SERVER"]["value"] = nfs_server
         env_by_name["NFS_PATH"]["value"] = nfs_mount
 
         nfs_volume = next(
-            v
-            for v in nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"]
-            if v.get("name") == "nfs-client-root"
+            (
+                v
+                for v in nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"]
+                if v.get("name") == "nfs-client-root"
+            ),
+            None,
         )
+        if nfs_volume is None:
+            raise ValueError("NFS deployment template missing volume 'nfs-client-root'")
+        if "nfs" not in nfs_volume:
+            raise ValueError(
+                "NFS deployment template volume 'nfs-client-root' is missing 'nfs' section"
+            )
         nfs_volume["nfs"]["server"] = nfs_server
         nfs_volume["nfs"]["path"] = nfs_mount
         log.info("Creating NFS client provisioner deployment")

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -10776,18 +10776,20 @@ def setup_nfs(request, setup_rbac):
         nfs_client_prov_dep_data = templating.load_yaml(
             constants.NFS_DEPLOYMENT_YAML_DIR
         )
-        nfs_client_prov_dep_data["spec"]["template"]["spec"]["containers"][0]["env"][1][
-            "value"
-        ] = nfs_server
-        nfs_client_prov_dep_data["spec"]["template"]["spec"]["containers"][0]["env"][2][
-            "value"
-        ] = nfs_mount
-        nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"][0]["nfs"][
-            "server"
-        ] = nfs_server
-        nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"][0]["nfs"][
-            "path"
-        ] = nfs_mount
+        container = nfs_client_prov_dep_data["spec"]["template"]["spec"]["containers"][
+            0
+        ]
+        env_by_name = {e.get("name"): e for e in container.get("env", [])}
+        env_by_name["NFS_SERVER"]["value"] = nfs_server
+        env_by_name["NFS_PATH"]["value"] = nfs_mount
+
+        nfs_volume = next(
+            v
+            for v in nfs_client_prov_dep_data["spec"]["template"]["spec"]["volumes"]
+            if v.get("name") == "nfs-client-root"
+        )
+        nfs_volume["nfs"]["server"] = nfs_server
+        nfs_volume["nfs"]["path"] = nfs_mount
         log.info("Creating NFS client provisioner deployment")
         create_resource(**nfs_client_prov_dep_data)
         nfs_client_dep_obj = OCS(**nfs_client_prov_dep_data)

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -5,6 +5,7 @@ import os
 import pandas as pd
 import random
 import secrets
+import socket
 import string
 import time
 import tempfile
@@ -10768,6 +10769,14 @@ def setup_nfs(request, setup_rbac):
         if not nfs_server or not nfs_mount:
             raise ValueError(
                 f"NFS config not set: nb_nfs_server={nfs_server}, nb_nfs_mount={nfs_mount}"
+            )
+
+        try:
+            socket.getaddrinfo(nfs_server, None, socket.AF_UNSPEC, socket.SOCK_STREAM)
+        except socket.gaierror:
+            pytest.skip(
+                f"NFS server '{nfs_server}' is not DNS-resolvable from this cluster; "
+                f"skipping NFS test"
             )
 
         log.info(

